lib/helpers.py

```diff
@@ -0,0 +1,389 @@
+        update_cache = True
+
+   
+    if df['user_id'].dtype != 'int64':
+      df = improve_types(df)
+      update_cache = True
+    
+    if update_cache:
+      print(datetime.now(), 'rewritting cached file with correct types (local and S3)', filename, s3_filename)
+      to_parquet(df, filename)
+      fs.put(filename, s3_filename)
+    
+    return df
+
+    
+def from_parquet(filename):
+    return pd.read_parquet(filename, engine='pyarrow')
+
+
+# helper to download CSV files, convert to DF and print time needed
+# caches files locally and on S3 to be reused
+def read_csv(customer, audience, source, date, columns=None, revenue_event=None, chunk_filter_fn=None, chunk_size=10 ** 6):
+    now = datetime.now()
+
+    date_str = date.strftime('%Y%m%d')
+    
+    filename = '{0}/{1}/{2}.csv.gz'.format(path(customer,audience), source, date_str)
+
+    # local cache
+    cache_dir = '{0}/{1}/{2}'.format(cache_folder, audience, source)
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+    
+    cache_filename = '{0}/{1}.parquet'.format(cache_dir, date_str)
+
+    # s3 cache (useful if we don't have enough space on the Colab instance)
+    s3_cache_filename = '{0}/{1}/{2}/{3}.parquet'.format(path(customer, audience),
+                                                           source, cache_folder, date_str)
+
+    if source == 'attributions':
+        cache_filename = '{0}/{1}-{2}.parquet'.format(cache_dir, date_str,
+                                                      revenue_event)
+
+        # s3 cache (useful if we don't have enough space on the Colab instance)
+        s3_cache_filename = '{0}/{1}/{2}/{3}-{4}.parquet' \
+            .format(path(customer, audience), source, cache_folder, date_str, revenue_event)
+
+    fs = s3fs.S3FileSystem(anon=False)
+    fs.connect_timeout = 10  # defaults to 5
+    fs.read_timeout = 30  # defaults to 15 
+
+    if os.path.exists(cache_filename):
+        print(now, 'loading from', cache_filename)
+        return from_parquet_corrected(cache_filename, s3_cache_filename, fs, columns)
+
+    
+    if fs.exists(path=s3_cache_filename):
+        print(now, 'loading from S3 cache', s3_cache_filename)
+        
+        # Download the file to local cache first to avoid timeouts during the load.
+        # This way, if they happen, restart will be using local copies first.
+        fs.get(s3_cache_filename, cache_filename)
+        
+        print(now, 'stored S3 cache file to local drive, loading', cache_filename)
+        
+        return from_parquet_corrected(cache_filename, s3_cache_filename, fs, columns)
+
+    print(now, 'start loading CSV for', audience, source, date)
+
+    read_csv_kwargs = {'chunksize': chunk_size}
+    if columns:
+        read_csv_kwargs['usecols'] = columns
+        
+    df = pd.DataFrame()
+    
+    if not fs.exists(path=filename):
+       print(now, 'WARNING: no CSV file at for: ', audience, source, date, ', skipping the file: ', filename)
```

Comment:
Maybe logging is more suitable here?

Reply:
Maybe later, when there's a class wrapping this thing up